vault/counters.go

```diff
@@ -0,0 +1,152 @@
+			if err != nil {
+				return nil, err
+			}
+
+			t, err := time.Parse(requestCounterDatePathFormat, fullpath)
+			if err != nil {
+				return nil, err
+			}
+
+			all.Dated = append(all.Dated, DatedRequestCounter{StartTime: t, RequestCounter: *counter})
+		}
+	}
+
+	return &all, nil
+}
+
+// loadCurrentRequestCounters reads the current RequestCounter out of storage.
+// The in-memory current request counter is populated with the value read, if any.
+// now should be the current time; it is a parameter to facilitate testing.
+func (c *Core) loadCurrentRequestCounters(ctx context.Context, now time.Time) error {
+	datepath := now.Format(requestCounterDatePathFormat)
+	counter, err := c.loadRequestCounters(ctx, datepath)
+	if err != nil {
+		return err
+	}
+	if counter != nil {
+		c.counters.activePath = datepath
+		atomic.StoreUint64(&c.counters.requests, *counter.Total)
+	}
+	return nil
+}
+
+// loadRequestCounters reads a RequestCounter out of storage at location datepath.
+// If nothing is found at that path, that isn't an error: a reference to a zero
+// RequestCounter is returned.
+func (c *Core) loadRequestCounters(ctx context.Context, datepath string) (*RequestCounter, error) {
+	view := c.systemBarrierView.SubView("counters/requests/")
+
+	out, err := view.Get(ctx, datepath)
+	if err != nil {
+		return nil, errwrap.Wrapf("failed to read request counters: {{err}}", err)
+	}
+	if out == nil {
+		return nil, nil
+	}
+
+	newCounters := &RequestCounter{}
+	err = out.DecodeJSON(newCounters)
+	if err != nil {
+		return nil, err
+	}
+
+	return newCounters, nil
+}
+
+// saveCurrentRequestCounters writes the current RequestCounter to storage.
+// The in-memory current request counter is reset to zero after writing if
+// we've entered a new month.
+// now should be the current time; it is a parameter to facilitate testing.
+func (c *Core) saveCurrentRequestCounters(ctx context.Context, now time.Time) error {
+	view := c.systemBarrierView.SubView("counters/requests/")
+	datepath := now.Format(requestCounterDatePathFormat)
+
+	var requests uint64
+	if c.counters.activePath == "" || c.counters.activePath == datepath {
+		// We leave requests at 0 in the case where the month has just changed; we don't
+		// want to write the current count of requests (from last month) to the new month
+		// datepath.  This means we discard any requests counted since the last time
+		// we wrote, but since we write frequently that's okay.
+		requests = atomic.LoadUint64(&c.counters.requests)
+	}
+
+	localCounters := &RequestCounter{
+		Total: &requests,
+	}
+	entry, err := logical.StorageEntryJSON(datepath, localCounters)
+	if err != nil {
+		return errwrap.Wrapf("failed to create request counters entry: {{err}}", err)
+	}
+
+	if err := view.Put(ctx, entry); err != nil {
+		return errwrap.Wrapf("failed to save request counters: {{err}}", err)
+	}
+
+	if datepath != c.counters.activePath {
+		if c.counters.activePath != "" {
```

Comment:
What scenario would activePath == ""?

Reply:
First save since Vault was started, when there was no existing saved value for the current month when we attempted to prime the counter at startup.